### PR TITLE
Implementation of the support for CKB trials

### PR DESCRIPTION
**Pull request by LiekeSchoenmaker**
SCOPE: 
- Solve INC-397 select highest evidence based on level and direction 
- Update TMB cutoff from 16 to 10
- Support the new serve data model (master branch of https://github.com/hartwigmedical/serve) 
- Update new data model of protect + loading into the database
- Add logica to support only NL clinical studies
- Add filter on gender of study 
- Sort ROSE output on alphabetically order 
- Update layout of the clinical evidence section of the rapportage 
- Update disclaimers of the patient reporter

### CODE REVIEW

File: common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java [hunk 1]
```diff
@@ -63,16 +67,48 @@ private static String header() {
...
+                .add("treatmentApproachesTherapy")
                 .add("onLabel")
                 .add("level")
                 .add("direction")
                 .add("sources")
                 .toString();
     }
 
+    @NotNull
+    private static String therapyName(@Nullable ClinicalTrial clinicalTrial, @Nullable Treatment treatment) {
```
Comment (mdenburger) [inline common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java:87]: The exact same method exists in `ProtectDAO.java`. Move to some common utility class and reuse the same method in both places?

File: common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java [hunk 2]
```diff
@@ -63,16 +67,48 @@ private static String header() {
...
+        boolean isClinicalTrial = clinicalTrial != null;
+        boolean isTreatment = treatment != null;
+
+        if (isClinicalTrial && isTreatment) {
+            throw new IllegalStateException("An actionable event cannot be both a treatment and clinical trial");
```
Comment (mdenburger) [inline common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java:92]: I'd log some details of the clinical trail and treatment here, so you can locate them sooner whenever this strange event happens

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                EvidenceDirection.RESPONSIVE), createProtectEvidence(EvidenceLevel.A, EvidenceDirection.PREDICTED_RESPONSIVE)));
+        assertFalse(ClinicalEvidenceFunctions.hasHigherOrEqualEvidenceForEventAndTrial(createProtectEvidenceList(EvidenceLevel.A,
+                EvidenceDirection.RESPONSIVE), createProtectEvidence(EvidenceLevel.A, EvidenceDirection.PREDICTED_RESPONSIVE)));
+        assertFalse(ClinicalEvidenceFunctions.hasHigherOrEqualEvidenceForEventAndTrial(createProtectEvidenceList(EvidenceLevel.A,
+                EvidenceDirection.RESPONSIVE), createProtectEvidence(EvidenceLevel.C, EvidenceDirection.RESISTANT)));
+
+    }
+
+    @NotNull
+    public List<ProtectEvidence> createProtectEvidenceList(@NotNull EvidenceLevel level, @NotNull EvidenceDirection direction) {
```
Comment (mdenburger) [inline patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java:52]: Can be `private`

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        List<ProtectEvidence> evidences = Lists.newArrayList();
+        evidences.add(createProtectEvidence(level, direction));
```
Comment (mdenburger) [inline patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java:54]: Can be simplified to `return Collections.singletonList(createProtectEvidence(level, direction))`. 

Whenever you need a list, set or map with only one item, use the `Collections.singletonXXX` methods: they're faster, simpler and cheaper to use and return an immutable collection.

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/ExampleAnalysisTestFactory.java [hunk 5]
```diff
@@ -342,16 +343,16 @@ private static List<ProtectEvidence> createCOLO829TumorSpecificEvidence() {
                 .reported(true)
                 .treatment(ImmutableTreatment.builder()
                         .name("Atezolizumab + Cobimetinib + Vemurafenib")
-                        .sourceRelevantTreatmentApproaches(Sets.newHashSet("RAF Inhibitor (Pan)",
+                        .treatmentApproachesDrugClass(Sets.newHashSet("RAF Inhibitor (Pan)",
                                 "MEK2 Inhibitor",
                                 "MEK inhibitor (Pan)",
                                 "MEK1 Inhibitor"))
-                        .relevantTreatmentApproaches(Sets.newHashSet())
+                        .treatmentApproachesTherapy(Sets.newHashSet())
```
Comment (mdenburger) [inline patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/ExampleAnalysisTestFactory.java:350]: Use `Collections.emptySet()` whenever you need an empty set. That's cheaper to create an always immutable. Same for lists and maps.

File: protect/src/main/java/com/hartwig/oncoact/protect/algo/EvidenceKey.java [hunk 6]
```diff
@@ -37,17 +37,17 @@ public static EvidenceKey create(@NotNull ProtectEvidence evidence) {
         return new EvidenceKey(evidence.gene(),
                 evidence.event(),
                 evidence.treatment().name(),
```
Comment (mdenburger) [inline protect/src/main/java/com/hartwig/oncoact/protect/algo/EvidenceKey.java:39]: According to the annotations, `treatment()` can be `null`. Something to account for here?

File: protect/src/main/java/com/hartwig/oncoact/protect/algo/ProtectAlgo.java [hunk 7]
```diff
@@ -179,12 +187,7 @@ public List<ProtectEvidence> run(@NotNull OrangeRecord orange) {
                 reportedCount(reported),
                 reportedCount(updatedForTrials));
 
-        List<ProtectEvidence> updatedForBlacklist = EvidenceReportingCuration.applyReportingBlacklist(updatedForTrials);
-        LOGGER.debug("Reduced reported evidence from {} items to {} items by blacklisting specific evidence for reporting",
-                reportedCount(updatedForTrials),
-                reportedCount(updatedForBlacklist));
-
-        return updatedForBlacklist;
+        return updatedForTrials;
```
Comment (mdenburger) [inline protect/src/main/java/com/hartwig/oncoact/protect/algo/ProtectAlgo.java:190]: No blacklist needed anymore for CKB?

File: protect/src/main/java/com/hartwig/oncoact/protect/evidence/CopyNumberEvidence.java [hunk 8]
```diff
@@ -55,20 +60,28 @@ public List<ProtectEvidence> evidence(@NotNull List<PurpleGainLoss> reportableSo
...
+    private List<ProtectEvidence> evidence(@NotNull PurpleGainLoss gainLoss, boolean report,
+            @Nullable PatientInformationResponse diagnosticPatientData) {
         List<ProtectEvidence> result = Lists.newArrayList();
         for (ActionableGene actionable : actionableGenes) {
             if (actionable.gene().equals(gainLoss.gene()) && isTypeMatch(actionable, gainLoss)) {
-                ProtectEvidence evidence = personalizedEvidenceFactory.somaticEvidence(actionable)
-                        .reported(report)
+                EvidenceType type = PersonalizedEvidenceFactory.determineEvidenceType(actionable, null);
+                if (type.equals(EvidenceType.PRESENCE_OF_PROTEIN) || type.equals(EvidenceType.ABSENCE_OF_PROTEIN)) {
+                    if (Genes.MSI_GENES.contains(actionable.gene())) {
```
Comment (mdenburger) [inline protect/src/main/java/com/hartwig/oncoact/protect/evidence/CopyNumberEvidence.java:77]: Can be simplified to `report = Genes.MSI_GENES.contains(actionable.gene())`

File: protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java [hunk 9]
```diff
@@ -90,33 +93,51 @@ private List<ProtectEvidence> evidence(@NotNull Variant variant) {
...
             }
         }
 
         for (ActionableGene gene : genes) {
             if (geneMatch(variant, gene)) {
-                evidences.add(evidence(variant, gene, mayReport && driverInterpretation == DriverInterpretation.HIGH, "gene"));
+                boolean report = mayReport && driverInterpretation == DriverInterpretation.HIGH;
+                EvidenceType type = PersonalizedEvidenceFactory.determineEvidenceType(gene, null);
+                if (type.equals(EvidenceType.ABSENCE_OF_PROTEIN)) {
+                    if (Genes.MSI_GENES.contains(gene.gene())) {
```
Comment (mdenburger) [inline protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java:125]: Can be simplified to `report = Genes.MSI_GENES.contains(gene.gene())`

File: protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java [hunk 10]
```diff
@@ -141,13 +162,12 @@ private ProtectEvidence evidence(@NotNull Variant variant, @NotNull ActionableEv
...
 
-        return personalizedEvidenceFactory.evidenceBuilderRange(actionable, range, rangeRank)
+        return personalizedEvidenceFactory.evidenceBuilderRange(actionable, range, rangeRank, diagnosticPatientData, report)
                 .gene(variant.gene())
                 .transcript(transcript)
                 .isCanonical(isCanonical)
                 .event(EventGenerator.variantEvent(variant))
                 .germline(isGermline)
-                .reported(report)
                 .eventIsHighDriver(driverInterpretation == null ? null : EvidenceDriverLikelihood.interpretVariant(driverInterpretation))
```
Comment (mdenburger) [inline protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java:171]: "reported" is no longer needed? It still exists in the builder.

File: protect/src/test/java/com/hartwig/oncoact/protect/evidence/VirusEvidenceTest.java [hunk 11]
```diff
@@ -35,7 +35,8 @@ public void canDetermineEvidenceForViruses() {
 
         VirusEvidence virusEvidence = new VirusEvidence(TestPersonalizedEvidenceFactory.create(), Lists.newArrayList(hpv, ebv));
 
-        List<ProtectEvidence> evidences = virusEvidence.evidence(testRecord);
+        List<ProtectEvidence> evidences = virusEvidence.evidence(testRecord, null);
+        ;
```
Comment (mdenburger) [inline protect/src/test/java/com/hartwig/oncoact/protect/evidence/VirusEvidenceTest.java:39]: Spurious comma? :)

**Review by mdenburger — CHANGES_REQUESTED** [review]

File: rose/README.md [hunk 12]
```diff
@@ -1,6 +1,7 @@
 # Relevant Oncogenic Summary of Eligibility
 
-ROSE makes an actionability summary of the clinical relevant (for the Netherlands) genomic events and signatures as determined by the
+ROSE makes an actionability summary of the clinical relevant (for the Netherlands) genomic events and signatures as
+determined by the
```
[the remark this answers is not in the document]
Comment (LiekeSchoenmaker) [reply rose/README.md:4]: updated

File: rose/src/main/java/com/hartwig/oncoact/rose/conclusion/ConclusionAlgo.java [hunk 13]
```diff
@@ -40,21 +25,20 @@
 import com.hartwig.oncoact.rose.ActionabilityConclusion;
 import com.hartwig.oncoact.rose.ImmutableActionabilityConclusion;
 import com.hartwig.oncoact.rose.RoseData;
-import com.hartwig.oncoact.rose.actionability.ActionabilityEntry;
-import com.hartwig.oncoact.rose.actionability.ActionabilityKey;
-import com.hartwig.oncoact.rose.actionability.Condition;
-import com.hartwig.oncoact.rose.actionability.ImmutableActionabilityKey;
-import com.hartwig.oncoact.rose.actionability.TypeAlteration;
+import com.hartwig.oncoact.rose.actionability.*;
```
[the remark this answers is not in the document]
Comment (LiekeSchoenmaker) [reply rose/src/main/java/com/hartwig/oncoact/rose/conclusion/ConclusionAlgo.java]: Have solved the problem already

File: pom.xml [hunk 14]
```diff
@@ -37,7 +37,7 @@
         <protect.version>3.0.1</protect.version>
         <rose.version>2.0.1</rose.version>
 
-        <serve.version>2.4.0</serve.version>
+        <serve.version>4.0.1-beta.1</serve.version>
```
[the remark this answers is not in the document]
Comment (LiekeSchoenmaker) [reply pom.xml]: I will move to formal release when serve verification is finished

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java (hunk 3 above)
Comment (LiekeSchoenmaker) [reply patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java:52]: updated

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/ExampleAnalysisTestFactory.java (hunk 5 above)
Comment (LiekeSchoenmaker) [reply patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/ExampleAnalysisTestFactory.java:350]: I have solved it in this class. However, this should be further incorporated into the whole project. I will updated when I see one

File: protect/src/main/java/com/hartwig/oncoact/protect/algo/ProtectAlgo.java (hunk 7 above)
Comment (LiekeSchoenmaker) [reply protect/src/main/java/com/hartwig/oncoact/protect/algo/ProtectAlgo.java:190]: Blacklisted is moved to SERVE, so I have removed it from this place

File: protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java (hunk 9 above)
Comment (LiekeSchoenmaker) [reply protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java:125]: done

File: protect/src/test/java/com/hartwig/oncoact/protect/evidence/VirusEvidenceTest.java (hunk 11 above)
Comment (LiekeSchoenmaker) [reply protect/src/test/java/com/hartwig/oncoact/protect/evidence/VirusEvidenceTest.java]: done

File: patient-reporter/src/main/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctions.java [hunk 15]
```diff
@@ -66,11 +78,12 @@ public static Map<String, List<ProtectEvidence>> buildTreatmentMap(@NotNull List
             if ((reportGermline || !evidence.germline()) && (requireOnLabel == null || evidence.onLabel() == requireOnLabel)) {
                 String treatment = Strings.EMPTY;
                 List<ProtectEvidence> treatmentEvidences = Lists.newArrayList();
-                Set<String> treatmentApproaches = evidence.treatment().sourceRelevantTreatmentApproaches();
+                Set<String> treatmentApproaches = extractCombinedTreatmentApproaches(evidence.treatment().treatmentApproachesDrugClass(),
```
[the remark this answers is not in the document]
Comment (LiekeSchoenmaker) [reply patient-reporter/src/main/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctions.java:81]: good catch, I have added an extra if statement

File: patient-reporter/src/main/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctions.java [hunk 16]
```diff
@@ -98,25 +111,73 @@ public static Map<String, List<ProtectEvidence>> buildTreatmentMap(@NotNull List
...
+            Boolean requireOnLabel, @NotNull String name) {
+        Map<String, List<ProtectEvidence>> evidencePerTreatmentMap = Maps.newHashMap();
+
+        for (ProtectEvidence evidence : evidences) {
+            if ((reportGermline || !evidence.germline()) && (requireOnLabel == null || evidence.onLabel() == requireOnLabel)) {
+                String trial = Strings.EMPTY;
+                List<ProtectEvidence> trialEvidences = Lists.newArrayList();
+                trial = evidence.clinicalTrial().studyTitle();
+                trialEvidences = evidencePerTreatmentMap.getOrDefault(trial, new ArrayList<>());
+                if (!hasHigherOrEqualEvidenceForEventAndTrial(trialEvidences, evidence) && !trial.equals(Strings.EMPTY)) {
```
[the remark this answers is not in the document]
Comment (LiekeSchoenmaker) [reply patient-reporter/src/main/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctions.java:125]: good point, I have implemented it

File: common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java (hunk 2 above)
Comment (LiekeSchoenmaker) [reply common/src/main/java/com/hartwig/oncoact/protect/ProtectEvidenceFile.java:92]: what should we log here? You want log a specific clinicalTrial or Treatment for every entry I assume?

File: patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java (hunk 4 above)
Comment (LiekeSchoenmaker) [reply patient-reporter/src/test/java/com/hartwig/oncoact/patientreporter/cfreport/chapters/analysed/ClinicalEvidenceFunctionsTest.java:54]: Update it. I will do it whenever I see it

File: protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java (hunk 10 above)
Comment (LiekeSchoenmaker) [reply protect/src/main/java/com/hartwig/oncoact/protect/evidence/VariantEvidence.java:167]: exists in PersonalizedEvidenceFactory.evidenceBuilder

File: common/src/main/java/com/hartwig/oncoact/protect/EvidenceComparator.java [hunk 17]
```diff
@@ -46,23 +50,35 @@ public int compare(@NotNull ProtectEvidence evidence1, @NotNull ProtectEvidence
...
         if (treatmentApproachesDrugClass != 0) {
             return treatmentApproachesDrugClass;
         }
 
-        int treatmentApproachesTherapy = evidence1.treatment()
-                .treatmentApproachesTherapy()
+        int treatmentApproachesTherapy = treatment1 != null && treatment2 != null ? treatment1.treatmentApproachesTherapy()
                 .toString()
-                .compareTo(evidence2.treatment().treatmentApproachesTherapy().toString());
+                .compareTo(treatment2.treatmentApproachesTherapy().toString()) : 0;
```
Comment (mdenburger) [inline common/src/main/java/com/hartwig/oncoact/protect/EvidenceComparator.java:81]: Many branches are not covered by unit tests. Maybe expand the `ProtectEvidenceComparatorTest` a bit?

File: common/src/main/java/com/hartwig/oncoact/util/ActionabilityIntervation.java [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+public class ActionabilityIntervation {
+
+    @NotNull
+    public static String therapyName(@Nullable ClinicalTrial clinicalTrial, @Nullable Treatment treatment) {
+        boolean isClinicalTrial = clinicalTrial != null;
+        boolean isTreatment = treatment != null;
+
+        if (isClinicalTrial && isTreatment) {
+            throw new IllegalStateException("An actionable event cannot be both a treatment and clinical trial");
```
Comment (mdenburger) [inline common/src/main/java/com/hartwig/oncoact/util/ActionabilityIntervation.java:20]: I'd include the clinical trail and treatment in the message to simplify debugging.

File: pom.xml [hunk 19]
```diff
@@ -37,7 +37,7 @@
         <protect.version>3.0.1</protect.version>
         <rose.version>2.0.1</rose.version>
 
-        <serve.version>local-SNAPSHOT</serve.version>
+        <serve.version>4.0.1-beta.2</serve.version>
```
Comment (mdenburger) [inline pom.xml:40]: Reminder to use a non-prerelease version before merging.

**Review by mdenburger — APPROVED** [review]
Approved with some non-blocking remarks.

**Review by mdenburger — APPROVED** [review]

**Pull request merged**